### PR TITLE
Dev cost management microsoft.cost management 2020 12 01 preview new

### PR DESCRIPTION
**Pull request by absyed**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [ ] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.exports.json [hunk 1]
```diff
@@ -0,0 +1,1136 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "version": "2020-12-01-preview",
```
Comment (jorgecotillo) [inline specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.exports.json:4]: This file only contains a version update, but the apis remain the same, I don't think this api needs to be included.

File: specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.json [hunk 2]
```diff
@@ -0,0 +1,2577 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "version": "2020-12-01-preview",
```
Comment (jorgecotillo) [inline specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.json:4]: This file only contains a version update, but the apis remain the same, I don't think this api needs to be included.

File: specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.generatedetailedcostreport.json [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+      }
+    },
+    "DownloadURL": {
+      "description": "The URL to download the generated report.",
+      "properties": {
+        "downloadUrl": {
+          "description": "The URL to download the generated report.",
+          "type": "string"
+        },
+        "validTill": {
```
Comment (jorgecotillo) [inline specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.generatedetailedcostreport.json:332]: perhaps consider a different name such as expirationDate, also might be worth adding the format as part of the description: e.g. 2020-10-01 01:01 UTC something like that.

File: specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.generatedetailedcostreport.json [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+          "enum": [
+            "ActualCost",
+            "AmortizedCost"
+          ],
+          "x-ms-enum": {
+            "name": "GenerateDetailedCostReportMetricType",
+            "modelAsString": true
+          }
+        },
+        "timePeriod": {
```
Comment (jorgecotillo) [inline specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.generatedetailedcostreport.json:236]: What happens when a POST operation has timePeriod and billingPeriod? which one wins? consider adding this info as part of the description.

File: specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.exports.json [hunk 5]
```diff
@@ -0,0 +1,1136 @@
...
+          "readOnly": true
+        },
+        "message": {
+          "description": "Error message indicating why the operation failed.",
+          "type": "string",
+          "readOnly": true
+        }
+      }
+    },
+    "Resource": {
```
Comment (yungezz) [inline specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.exports.json:684]: Curious why all properties are readOnly?

**Comment by yungezz** [comment]
Need ARM review since new API version added. hi @absyed, could you pls fix LintDiff CI failure?

File: specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.exports.json (hunk 1 above)
Comment (absyed) [reply specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.exports.json]: Removed it

File: specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.json (hunk 2 above)
Comment (absyed) [reply specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.json]: Removed it

File: specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.generatedetailedcostreport.json (hunk 3 above)
Comment (absyed) [reply specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.generatedetailedcostreport.json:336]: Added more information in description

File: specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.generatedetailedcostreport.json (hunk 4 above)
Comment (absyed) [reply specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.generatedetailedcostreport.json:240]: If the customer has both timePeriod and billingPeriod, API will return back a 400 and ask the client to only use one. This is what I have in description for them, let me know if I should update it "Can only have one of either timePeriod or invoiceId or billingPeriod parameters."

File: specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.exports.json (hunk 5 above)
Comment (absyed) [reply specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.exports.json:684]: Keeping it consistent with how it already has been defined in costmangement.json

**Comment by jorgecotillo** [comment]
ARM signed off

File: specification/cost-management/resource-manager/readme.md [hunk 6]
```diff
@@ -32,6 +32,16 @@ azure-validator: false
 
 ---
 
+### Tag: package-preview-2020-12
+
+These settings apply only when `--tag=package-preview-2020-12` is specified on the command line.
+
+```yaml $(tag) == 'package-preview-2020-12'
```
Comment (yungezz) [inline specification/cost-management/resource-manager/readme.md:39]: why this preview tag contains table version Swaggers?

**Comment by yungezz** [comment]
hi @msyyc could you pls check impact to Python SDK? thanks

**Comment by msyyc** [comment]
> hi @msyyc could you pls check impact to Python SDK? thanks

It's ok for Python

**Comment by absyed** [comment]
> > hi @msyyc could you pls check impact to Python SDK? thanks
> 
> It's ok for Python

@yungezz anything else missing here

**Comment by yungezz** [comment]
> > > hi @msyyc could you pls check impact to Python SDK? thanks
> > 
> > 
> > It's ok for Python
> 
> @yungezz anything else missing here

HI @absyed, is service side change on Prod now? If yes, the PR is ok to merge. I'll remove `donotmerge` label then.

**Comment by msyyc** [comment]
For python, the breaking change is not caused by this PR

**Comment by absyed** [comment]
> > > > hi @msyyc could you pls check impact to Python SDK? thanks
> > > 
> > > 
> > > It's ok for Python
> > 
> > 
> > @yungezz anything else missing here
> 
> HI @absyed, is service side change on Prod now? If yes, the PR is ok to merge. I'll remove `donotmerge` label then.

Hi @yungezz, server side changes are in Production now and the PR is Ok to be merged in.

File: specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.generatedetailedcostreport.json [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+            "headers": {
+              "Location": {
+                "description": "The URL to check the result of the asynchronous operation.",
+                "type": "string"
+              },
+              "Azure-Consumption-AsyncOperation": {
+                "description": "The URL to check the status of the asynchronous operation.",
+                "type": "string"
+              },
+              "Azure-AsyncOperation": {
```
Comment (yungezz) [inline specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.generatedetailedcostreport.json:102]: what's difference between `Azure-AsyncOperation` and `Azure-Consumption-AsyncOperation`?

**Review by yungezz — APPROVED** [review]

**Comment by yungezz** [comment]
hi @absyed, before merging, one thing to confirm, about the swagger file name `costmanagement.generatedetailedcostreport.json `, why it's naming as `xxx.xxx.json`, since normal swagger file is just `<resourcename>.json`

**Comment by absyed** [comment]
> hi @absyed, before merging, one thing to confirm, about the swagger file name `costmanagement.generatedetailedcostreport.json `, why it's naming as `xxx.xxx.json`, since normal swagger file is just `<resourcename>.json`. No idea of impact of it, I saw cost management has swagger file name in that format already

Hi @yungezz, this is following the  costmanagement.export.json which already exist. In this new preview API version is the first time generatedetailedcostreport API is being introduced.

File: specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.generatedetailedcostreport.json (hunk 7 above)
Comment (absyed) [reply specification/cost-management/resource-manager/Microsoft.CostManagement/preview/2020-12-01-preview/costmanagement.generatedetailedcostreport.json:102]: The both return the same value.
We have been offering both headers for this service for some time, so am maintaining consistency here.

**Comment by absyed** [comment]
@lirenhe can you please let me know if there is anything else required for this PR to be checked in?

**Review by yungezz — APPROVED** [review]

**Pull request merged**